File: WasteVisionWebBE/Domain/Prediction/PredictionMapper.cs
using System.Collections.Generic;
using System.Linq;
using DDDSample1.Domain.Users;
using DDDSample1.Domain.RoboflowModels;
using DDDSample1.Domain.Shared;

namespace DDDSample1.Domain.Predictions
{
    public static class PredictionMapper
    {
        public static PredictionDto ToDto(Prediction prediction)
        {
            if (prediction == null) return null;

            return new PredictionDto(
                prediction.Id.AsGuid(),
                prediction.UserId.AsString(),
                prediction.RoboflowModelId.AsString(),
                prediction.OriginalImageUrl.AsString(),
                prediction.ProcessedImageUrl.AsString(),
                prediction.Date.AsString() // ISO 8601 format
            );
        }

         public static PredictionApiDto ToApiDto(Prediction prediction)
        {
            if (prediction == null) return null;

            return new PredictionApiDto(
                prediction.Id.AsGuid(),
                prediction.UserId.AsString(),
                prediction.User.Username.AsString(),
                prediction.RoboflowModelId.AsString(),
                prediction.RoboflowModel.Description.AsString(),
                prediction.OriginalImageUrl.AsString(),
                prediction.ProcessedImageUrl.AsString(),
                prediction.Date.AsString() // ISO 8601 format
            );
        }

        public static List<PredictionDto> ToDtoList(List<Prediction> predictions)
        {
            if (predictions == null) return null;
            return predictions.Select(pred => ToDto(pred)).ToList();
        }

          public static List<PredictionApiDto> ToApiDtoList(List<Prediction> predictions)
        {
            if (predictions == null) return null;
            return predictions.Select(pred => ToApiDto(pred)).ToList();
        }

        public static Prediction ToDomain(CreatingPredictionDto dto)
        {
            if (dto == null) return null;

            return new Prediction(
                new UserId(dto.UserId),
                new RoboflowModelId(dto.ModelId),
                new Url(dto.OriginalImageUrl),
                new Url(dto.ProcessedImageUrl)
            );
        }
    }
}

File: WasteVisionWebFE/src/data/mappers/prediction-mapper.ts
import { Prediction } from "@/core/domain/Prediction";
import { PredictionDTO } from "../dto/prediction-dto";

export class PredictionMapper {
  /**
   * Maps a PredictionDTO to a domain Prediction
   */
  public static toDomain(dto: PredictionDTO): Prediction {
    return {
      id: dto.id,
      userId: dto.userId,
      modelId: dto.modelId,
      originalImageUrl: dto.originalImageUrl,
      processedImageUrl: dto.processedImageUrl,
      roboflowModel: {
        id: dto.modelId,
        description: dto.modelName || "",
        apiKey: "",
        modelUrl: "",
        localModelPath: "",
        endPoint: "",
        map: "",
        recall: "",
        precision: "",
        active: true,
      },
      user:{
        id: dto.userId,
        username: dto.userName || "",
        email: "",
        active: true,
      },
      date: dto.date,
    };
  }


  /**
   * Maps a domain Prediction to a PredictionDTO
   */
  public static toDTO(domain: Prediction): PredictionDTO {
    return {
      id: domain.id,
      userId: domain.userId,
      modelId: domain.modelId,
      originalImageUrl: domain.originalImageUrl,
      processedImageUrl: domain.processedImageUrl,
      date: domain.date,
      roboflowModel: domain.roboflowModel,
      user: domain.user,
    };
  }

  /**
   * Maps an array of PredictionDTOs to domain Predictions
   */
  public static toDomainList(dtos: PredictionDTO[]): Prediction[] {
    return dtos.map((dto) => this.toDomain(dto));
  }


}